debian/man/ascheck.1
.\" Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.\" Please adjust this date whenever revising the manpage.
.\" 
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.TH "ASCHECK" "1" "June 3, 2025" "" ""
.SH "NAME"
ascheck \- check integrity of EPICS access security files
.SH "SYNOPSIS"
.B ascheck
[-S\ "mac=sub,..."] [< filename]

.SH "DESCRIPTION"
.B ascheck
displays errors on stdout
when a file is redirected to it.
If no errors are detected it prints nothing.
Only syntax errors, not logic errors, are detected.
Thus it is still possible to get yourself in trouble.
The flag -S means a set of macro substitutions may appear.
This is just like the macro substitutions for dbLoadDatabase.

.SH "OPTIONS"
.TP
.B -S
This option allows macro substitutions to be used in the file.
It is a single string of the form
.B "mac1=sub1,mac2=sub2,..."


debian/changelog
epics-base (3.14.11-9~bpo50+1) lenny-backports; urgency=low

  * Rebuild for lenny-backports.
  * switch to readline 5

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Sun, 11 Sep 2011 14:26:43 -0400

epics-base (3.14.11-9) unstable; urgency=low

  [ Till Straumann ]
  * RTEMS: Time registration when dynamically loading
     Fixes lp:831648
  * libCom: Fix OS priority mapping on Posix
     Fixes lp:835138

  [ Andrew Johnson ]
  * shutdown: Lock records during dbCa link cleanup

  [ Michael Davidsaver ]
  * libCom: Avoid race in errlog shutdown.
  * linux: Use RT thread priorities if available
  * Set EPICS_SITE_VERSION to debian package revision

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Sun, 11 Sep 2011 14:21:22 -0400

epics-base (3.14.11-8) unstable; urgency=low

  [ Michael Davidsaver ]
  * Fix postinst
  * Change default max array size to 1M
  * move conf.d to CONFIG

  [ Nick Hastings ]
  * g++-4.6 fix: "#include <cstddef>"

  [ Michael Davidsaver ]
  * override CONFIG
  * Use 3.14.12 method to detect EPICS_HOST_ARCH if unset
  * test of timerQ ref counting

  [ Jeff Hill ]
  * only delete a timer queue when the ref count is 0
     Fixes LP 786979

  [ Michael Davidsaver ]
  * Drop "mutable" references
     Fixes LP 736273
  * Missing bspExt for mvme5500
  * Install sample Makefile
  * correct location for EpicsHostArch.pl
  * pcas headers use install list

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Mon, 06 Jun 2011 10:29:06 -0400

epics-base (3.14.11-7) unstable; urgency=low

  * libCom: Fix errlogPrintf() assert fail at shutdown
  * fix description of -x in softIoc manpage
  * detect EPICS_HOST_ARCH if unset
  * use auto host detection
  * manpage for softIoc in wrong package
  * relocate configure/ as /etc/epics/configure
  * populate ALL_CROSS_COMPILER_TARGET_ARCHS with debug target

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Thu, 12 May 2011 17:38:15 -0400

epics-base (3.14.11-6) unstable; urgency=low

  * change readline to 6

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Tue, 16 Nov 2010 10:45:47 -0500

epics-base (3.14.11-5) unstable; urgency=low

  * incorrect path in S99caRepeater and S99logServer init scripts
      Reported by Martin Konrad

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Wed, 10 Nov 2010 11:01:06 -0500

epics-base (3.14.11-4) unstable; urgency=low

  * fix use of temp files during builds
      Fixes problem with parallel builds expanding MSI templates
  * rtems: Bind NTP client to a random port
      Fixes time sync bug on networks with NTP broadcasts
      Launchpad bug #663875

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Wed, 20 Oct 2010 13:06:50 -0400

epics-base (3.14.11-3) unstable; urgency=low

  * wrong paths in perl scripts
  * correct dbd path for softIoc

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Wed, 08 Sep 2010 15:02:53 -0400

epics-base (3.14.11-2) unstable; urgency=low

  * change install location to /usr/lib/epics/
  * add manpages for makeBaseApp/Ext and softIoc
  * update manpages for caget, camonitor, and caput
  * update lintian overrides

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Mon, 16 Aug 2010 13:54:32 -0400

epics-base (3.14.11-1) unstable; urgency=low

  * New upstream version
  * Split RTEMS packages to separate source package rtems-epics
  * Applied Known Problems patches: softIoc, long_string, and macCore
  * Backported from bazaar repository
      fix for mantis 366
      ESHUTDOWN is not in FreeBSD and hence not in RTEMS.
      J field should not be PP, to match others.
      Fix UDF processing so we respect IVOA.
      Fixed crash when ALG (algorithm) was changed 
        to Average at runtime. (Fixes #552803)
      Fix NULL handle issue in macParseDefns()

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Thu, 15 Jul 2010 14:27:29 -0400

epics-base (3.14.10-16) unstable; urgency=low

  * Forgot device() entry for historgram record
  * Remove caRepeater init script.  It isn't working correctly.

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Mon, 12 Apr 2010 23:14:19 -0400

epics-base (3.14.10-15) unstable; urgency=low

  * Typo in caRepeater script

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Mon, 12 Apr 2010 13:21:15 -0400

epics-base (3.14.10-14) unstable; urgency=low

  * init script for caRepeater (from IIT)
  * base.dbd: include histogram record
  * debian/rules: use dh
  * build debug package libepics3.14.10-dbg
  * epics-dev: remove /usr/epics/base/bin/host symlink
  * epics-perl: move /usr/epics/base/lib/host symlink
  * patches:
     rec: initialize all old/prev./last value fields

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Fri, 26 Mar 2010 13:25:29 -0400

epics-base (3.14.10-13) unstable; urgency=low

  * osdWireConfig.h: Replace C++ style comments with C style
  * Add manpages for caget, caput, camonitor, cainfo, and caRepeater
    (from Ken McIvor <mcivor@iit.edu> and Carlo Segre <segre@debian.org>)
  * RTEMS: Add rtemsCom to the list of libraries to exclude from
    GeSys loadable objects.
  * boRecord: initialize the mlst field so that monitors are posted 
    correctly

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Wed, 24 Mar 2010 11:18:18 -0400

epics-base (3.14.10-12) unstable; urgency=low

  * Build for RTEMS-mvme2307

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Wed, 17 Feb 2010 09:32:52 -0500

epics-base (3.14.10-11) unstable; urgency=low

  * Recommend: msi
  * Depends on RTEMS 4.9.3

 -- Michael Davidsaver <mdavidsaver@bnl.gov>  Tue, 16 Feb 2010 16:54:34 -0500

epics-base (3.14.10-10) unstable; urgency=low

  * Rename rtems packages to follow convention (rtems-modname-bsp)
  * Rename epics-libs to libepics3.14.10 to follow library
    package naming convention.
  * Symlink to makeBaseApp, makeBaseExt, and softIoc in usr/bin/
  * move libCap5 to usr/epics/base/lib/linux-x86 to avoid
    potential conflicts with user compiled versions
  * For bin and lib create a symlink usr/epics/base/*/host
    to the host target name.


src/as/ascheck.c
/*************************************************************************\
* Copyright (c) 2002 The University of Chicago, as Operator of Argonne
*     National Laboratory.
* Copyright (c) 2002 The Regents of the University of California, as
*     Operator of Los Alamos National Laboratory.
* EPICS BASE Versions 3.13.7
* and higher are distributed subject to a Software License Agreement found
* in file LICENSE that is included with this distribution. 
\*************************************************************************/

/* ascheck.c,v 1.9.2.3 2006/12/01 23:52:59 anj Exp */
/*	Author: Marty Kraimer	Date: 03-24-94	*/

#include <stdlib.h>
#include <stddef.h>
#include <stdio.h>
#include <string.h>

#include "errlog.h"
#include "asLib.h"
#include "dbStaticLib.h"
int main(int argc,char **argv)
{
    int         argn = 1;
    char	*sub = NULL;
    int		subLength = 0;
    char	**pstr;
    char	*psep;
    int		*len;
    long	status = 0;
    static char *subSep = ",";

    /* Look for -Smacro=value options */
    while (argc>argn && (strncmp(argv[argn], "-S", 2)==0)) {
	pstr = &sub;
	psep = subSep;
	len = &subLength;
	if (strlen(argv[argn])==2) {
	    dbCatString(pstr, len, argv[++argn], psep);
	} else {
	    dbCatString(pstr, len, argv[argn]+2, psep);
	}
	argn++;
    }
    if (argc == argn) {
	status = asInitFP(stdin, sub);
	if(status) errlogPrintf("ascheck: Access Security File failed.\n");
    } else if (argc == argn+1) {
	status = asInitFile(argv[argn], sub);
	if(status) errlogPrintf("ascheck: Access Security File failed.\n");
    } else {
	printf("usage: ascheck [-Smac=sub ...] [<] file\n");
	status = -1;
    }
    errlogFlush();
    return status;
}
